### ISSUE TITLE
Investigate Fannkuch_9 failures on ARM64

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/performance#1483]
I have disabled this test for now so that we don't lose 600+ results per run, but we need to figure out why this test is failing.

**Comment by adamsitnik** [comment]
@DrewScoggins do you have any logs from the failure? I've tried to repro it on my ARM64 machine (Surface Pro X) and it works fine. I am asking because if there is a bug in dotnet/runtime we should report it and fix asap as well (RC1 + the code of the benchmark itself is not using any vector magic and it should just work)

**Comment by DrewScoggins** [comment]
I have posted the log below, you can get the full log here, https://helix.dot.net/api/2019-06-17/jobs/5338cd65-e862-49cf-bd4c-84c440dd1d4b/workitems/Partition4/console?access_token={Get from helix.dot.net}

I agree we should try and get this figured out, but we need to unblock the lab as well.

```
[2020/08/25 02:52:10][INFO] OverheadJitting  1: 1 op, 671543.00 ns, 671.5430 us/op
[2020/08/25 02:52:10][INFO] 
[2020/08/25 02:52:10][INFO] System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
[2020/08/25 02:52:10][INFO]  ---> System.Exception: Expected 556355 actual 556375
[2020/08/25 02:52:10][INFO]    at BenchmarksGame.FannkuchRedux_9.RunBench(Int32 n, Int32 expectedSum) in /home/helixbot/work/A4F008AB/w/A835093E/e/src/benchmarks/micro/runtime/BenchmarksGame/fannkuch-redux-9.cs:line 149
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Autogenerated.Runnable_8.WorkloadActionNoUnroll(Int64 invokeCount) in /home/helixbot/work/A4F008AB/w/A835093E/e/artifacts/bin/MicroBenchmarks/Release/netcoreapp3.1/ff2bed35-119d-46c3-904e-e18ce54ccc5c/ff2bed35-119d-46c3-904e-e18ce54ccc5c.notcs:line 7008
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Engines.Engine.RunIteration(IterationData data)
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Engines.EngineFactory.Jit(Engine engine, Int32 jitIndex, Int32 invokeCount, Int32 unrollFactor)
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Engines.EngineFactory.CreateReadyToRun(EngineParameters engineParameters)
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Autogenerated.Runnable_8.Run(IHost host, String benchmarkName) in /home/helixbot/work/A4F008AB/w/A835093E/e/artifacts/bin/MicroBenchmarks/Release/netcoreapp3.1/ff2bed35-119d-46c3-904e-e18ce54ccc5c/ff2bed35-119d-46c3-904e-e18ce54ccc5c.notcs:line 6679
[2020/08/25 02:52:10][INFO]    --- End of inner exception stack trace ---
[2020/08/25 02:52:10][INFO]    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
[2020/08/25 02:52:10][INFO]    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
[2020/08/25 02:52:10][INFO]    at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
[2020/08/25 02:52:10][INFO]    at BenchmarkDotNet.Autogenerated.UniqueProgramName.AfterAssemblyLoadingAttached(String[] args) in /home/helixbot/work/A4F008AB/w/A835093E/e/artifacts/bin/MicroBenchmarks/Release/netcoreapp3.1/ff2bed35-119d-46c3-904e-e18ce54ccc5c/ff2bed35-119d-46c3-904e-e18ce54ccc5c.notcs:line 671
[2020/08/25 02:52:10][INFO] // AfterAll
```

**Comment by adamsitnik** [comment]
@kunalspathak @AndyAyersMS could this be an ARM64 only bug introduced to the product recently?

**Comment by kunalspathak** [comment]
I will try to reproduce it. Does it happen on any specific platform like Windows or Linux only or doesn't matter. Also, could you please paste the command line parameters to reproduce it?

**Comment by adamsitnik** [comment]
> Also, could you please paste the command line parameters to reproduce it?

```cmd
git clone https://github.com/dotnet/performance.git
py .\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter FannkuchRedux_9 --architecture arm64
```

 I am not able to reproduce it on Windows (`Surface Pro X`). The python script downloads the latest SDK and for me, the Runtime version was `5.0.20.41714` when I was checking it few minutes. I wonder whether this a Linux specific issue, it's related to number of CPUs (the [code is dependent on it](https://github.com/dotnet/performance/blob/ac44a9b14552c289e10ad594c56a56492f456f66/src/benchmarks/micro/runtime/BenchmarksGame/fannkuch-redux-9.cs#L163-L164)) or it is a recent bug that is not part of SDK yet (the failure comes from runtime repo where a local build of dotnet runtime is used)

> Does it happen on any specific platform like Windows or Linux only or doesn't matter. 

From what I know the lab uses Ubuntu 18.04. @DrewScoggins please correct me if I am wrong

**Comment by kunalspathak** [comment]
I am able to reproduce this with `CoreRun.exe` of https://github.com/dotnet/runtime/commit/c9da1ed4ad677ed35d7bf9402de60c9b7d6d4a6e which is from last week and `dotnet` with rc1. Here is the output on my machine with `dotnet --list-sdks`
```
5.0.100-preview.3.20169.11 [c:\kpathak\git\performance\tools\dotnet\arm64\sdk]
5.0.100-preview.6.20270.13 [c:\kpathak\git\performance\tools\dotnet\arm64\sdk]
5.0.100-preview.7.20317.1 [c:\kpathak\git\performance\tools\dotnet\arm64\sdk]
5.0.100-preview.7.20317.11 [c:\kpathak\git\performance\tools\dotnet\arm64\sdk]
5.0.100-rc.1.20413.9 [c:\kpathak\git\performance\tools\dotnet\arm64\sdk]
```

**Comment by DrewScoggins** [comment]
I am going to give this a run on some of our lab machines as well. Would a core dump be useful?

**Comment by AndyAyersMS** [comment]
> could this be an ARM64 only bug introduced to the product recently

@adamsitnik could be, yes. 

@kunalspathak since you can repro, can you own root causing this? It will likely repro with builds from master too.

**Comment by DrewScoggins** [comment]
I also reproduced this on `.NET Core SDK=6.0.100-alpha.1.20426.4`, so you should be able to just use master.

**Comment by kunalspathak** [comment]
So the test spawns threads equal to no. of processors to do a specific task. The test fails for certain no. of threads. Here are the list of threads for it fails. The machine I am testing has 46 cores. I created a standalone application to repro it.

```
Expected 556355 actual -149195 for threads: 13
Expected 556355 actual 89389 for threads: 17
Expected 556355 actual 556357 for threads: 19
Expected 556355 actual -74053 for threads: 23
Expected 556355 actual -256488 for threads: 26
Expected 556355 actual 49252 for threads: 29
Expected 556355 actual 556390 for threads: 31
Expected 556355 actual 67096 for threads: 34
Expected 556355 actual 556358 for threads: 37
Expected 556355 actual 556357 for threads: 38
Expected 556355 actual -109454 for threads: 39
Expected 556355 actual 556374 for threads: 41
Expected 556355 actual 11077 for threads: 43
Expected 556355 actual 556375 for threads: 46
```

I am wondering if this ever passed on ARM64? When was the last time it passed because I see this test recently added by @danmosemsft  in https://github.com/dotnet/performance/pull/1453 ?

**Comment by AndyAyersMS** [comment]
Maybe the algorithm is buggy an only works for certain numbers of threads?

**Comment by kunalspathak** [comment]
> Maybe the algorithm is buggy an only works for certain numbers of threads?

Yes, that's what it is. On my dev machine, I get exact same output when tried no. of threads 1 thru 46. Although I just have 12 cores, the first failure happens on 13 threads.

Here is the `dotnet --list-sdks` output on my dev machine:

```
3.1.400-preview-015151 [C:\Program Files\dotnet\sdk]
3.1.401 [C:\Program Files\dotnet\sdk]
5.0.100-preview.7.20317.11 [C:\Program Files\dotnet\sdk]
5.0.100-preview.7.20366.6 [C:\Program Files\dotnet\sdk]
```

FYI @danmosemsft .

**Comment by kunalspathak** [comment]
Here is the standalone code, if anyone is interested.

<details>
<summary>Repro code</summary>

```cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

// Adapted from fannkuch-redux C# .NET Core #9 program
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/fannkuchredux-csharpcore-9.html
// Best-scoring single-threaded C# .NET Core version as of 2020-08-13

// The Computer Language Benchmarks Game
// https://benchmarksgame-team.pages.debian.net/benchmarksgame/
//
// contributed by Flim Nik
// small optimisations by Anthony Lloyd

using System;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
// using BenchmarkDotNet.Attributes;
// using MicroBenchmarks;

namespace BenchmarksGame
{
    // [BenchmarkCategory(Categories.Runtime, Categories.BenchmarksGame, Categories.JIT)]
    public unsafe class FannkuchRedux_9
    {
        static int taskCount;
        static int[] fact, chkSums, maxFlips;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        static void FirstPermutation(short* p, short* pp, int* count, int n, int idx)
        {
            for (int i = 0; i < n; ++i) p[i] = (byte)i;
            for (int i = n - 1; i > 0; --i)
            {
                int d = idx / fact[i];
                count[i] = d;
                if (d > 0)
                {
                    idx %= fact[i];
                    for (int j = i; j >= 0; --j) pp[j] = p[j];
                    for (int j = 0; j <= i; ++j) p[j] = pp[(j + d) % (i + 1)];
                }
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        static void NextPermutation(short* p, int* count)
        {
            var first = p[1];
            p[1] = p[0];
            p[0] = first;
            int i = 1;
            while (++count[i] > i)
            {
                count[i++] = 0;
                var next = p[1];
                p[0] = next;
                for (int j = 1; j < i;) p[j] = p[++j];
                p[i] = first;
                first = next;
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        static void Copy(short* p, short* pp, int n)
        {
            var startL = (long*)p;
            var stateL = (long*)pp;
            var lengthL = n / 4;
            int i = 0;
            for (; i < lengthL; i++)
            {
                stateL[i] = startL[i];
            }
            for (i = lengthL * 4; i < n; i++)
            {
                pp[i] = p[i];
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        static int CountFlips(short* p, short* pp, int n)
        {
            int flips = 1;
            int first = *p;
            short temp;
            if (p[first] != 0)
            {
                Copy(p, pp, n);
                do
                {
                    ++flips;
                    if (first > 2)
                    {
                        short* lo = pp + 1, hi = pp + first - 1;
                        do
                        {
                            temp = *lo;
                            *lo = *hi;
                            *hi = temp;
                        } while (++lo < --hi);
                    }
                    temp = pp[first];
                    pp[first] = (short)first;
                    first = temp;
                } while (pp[first] != 0);
            }
            return flips;
        }

        static void Run(int n, int taskSize)
        {
            int* count = stackalloc int[n];
            int taskId, chksum = 0, maxflips = 0;
            short* p = stackalloc short[n];
            short* pp = stackalloc short[n];
            while ((taskId = Interlocked.Decrement(ref taskCount)) >= 0)
            {
                FirstPermutation(p, pp, count, n, taskId * taskSize);
                if (*p != 0)
                {
                    var flips = CountFlips(p, pp, n);
                    chksum += flips;
                    if (flips > maxflips) maxflips = flips;
                }
                for (int i = 1; i < taskSize; i++)
                {
                    NextPermutation(p, count);
                    if (*p != 0)
                    {
                        var flips = CountFlips(p, pp, n);
                        chksum += (1 - (i & 1) * 2) * flips;
                        if (flips > maxflips) maxflips = flips;
                    }
                }
            }
            chkSums[-taskId - 1] = chksum;
            maxFlips[-taskId - 1] = maxflips;
        }

        public static void Main(string[] args)
        {
            FannkuchRedux_9 a = new FannkuchRedux_9();
            a.RunBench(11, 556355);
        }

        // Official runs use [Arguments(12, 3968050)] which takes ~4.2 sec vs ~330ms for 11
        // [Benchmark(Description = nameof(FannkuchRedux_9))]
        // [Arguments(11, 556355)]
        public int RunBench(int n, int expectedSum)
        {
            int chkSum = 0;
            for (int numThread = 1; numThread <= 46; numThread++)
            {
                chkSum = Bench(n, false, numThread);
                if (chkSum != expectedSum)
                {
                    // throw new Exception($"Expected {expectedSum} actual {chkSum} for {numThread}");
                    Console.WriteLine($"Expected {expectedSum} actual {chkSum} for threads: {numThread}");
                }
            }
            return chkSum;
        }

        public static int Bench(int n, bool verbose, int PC)
        {
            fact = new int[n + 1];
            fact[0] = 1;

            for (int i = 1; i < fact.Length; i++)
            {
                fact[i] = fact[i - 1] * i;
            }

            // var PC = 46;//Environment.ProcessorCount;
            // Console.WriteLine("PC: " + PC);
            taskCount = n > 11 ? fact[n] / (9 * 8 * 7 * 6 * 5 * 4 * 3 * 2) : PC;
            int taskSize = fact[n] / taskCount;
            chkSums = new int[PC];
            maxFlips = new int[PC];
            var threads = new Thread[PC];
            for (int i = 1; i < PC; i++)
            {
                (threads[i] = new Thread(() => Run(n, taskSize))).Start();
            }
            Run(n, taskSize);

            for (int i = 1; i < threads.Length; i++)
            {
                threads[i].Join();
            }
            int chkSum = chkSums.Sum();
            if (verbose) Console.WriteLine(chkSum + "\nPfannkuchen(" + n + ") = " + maxFlips.Max());

            return chkSum;
        }
    }
}
```
</details>

**Comment by AndyAyersMS** [comment]
This bit seems suspect: 
```
            taskCount = n > 11 ? fact[n] / (9 * 8 * 7 * 6 * 5 * 4 * 3 * 2) : PC;
```
seems like `taskCount` should evenly divide `fact[n]`?

Looks like failing cases are all numbers that have a prime > 11 in their factorization.

**Comment by adamsitnik** [comment]
@kunalspathak @AndyAyersMS big thanks for your help and apologies for the false alarm. I've fixed it via #1488
